packages/tpetra/core/example/BlockCrs/driver.cpp

```diff
@@ -0,0 +1,536 @@
+            
+            auto block = blocks(loc);
+            for (LO l0=0;l0<blocksize;++l0) 
+              for (LO l1=0;l1<blocksize;++l1) 
+                block(l0, l1) = get_block_crs_entry<value_type>(i0, j0, k0,
+                                                                diff_i, diff_j, diff_k,
+                                                                l0, l1);
+          }
+        });
+    }
+    {
+      RCP<TimeMonitor> timerBlockCrsFillComplete 
+        = rcp(new TimeMonitor(*TimeMonitor::getNewTimer("3) BlockCrsMatrix FillComplete - currently do nothing")));
+      // this function does not exsit 
+      //A_bcrs->fillComplete();
+    }
+#if PRINT_VERBOSE
+    A_bcrs->describe(*out, Teuchos::VERB_EXTREME);
+#endif
+
+    // Create MultiVector
+    RCP<tpetra_multivector_type> X = Teuchos::rcp(new tpetra_multivector_type(A_bcrs->getDomainMap(), nrhs));
+    {
+      RCP<TimeMonitor> timerMultiVectorFill
+        = rcp(new TimeMonitor(*TimeMonitor::getNewTimer("4) MultiVectorFill")));
+      
+      auto value = X->getLocalView<typename exec_space::memory_space>();
+      auto map = X->getMap()->getLocalMap();
+      Kokkos::parallel_for
+        (value.extent(0), KOKKOS_LAMBDA(const LO i) {
+          const GO gid = map.getGlobalElement(i);
+          for (LO j=0,jend=value.extent(1);j<jend;++j) 
+            value(i, j) = get_multi_vector_entry<value_type>(gid, j);
+        });
+    }
+    
+#if PRINT_VERBOSE
+    X->describe(*out, Teuchos::VERB_EXTREME);
+#endif
+
+    RCP<tpetra_multivector_type> B_bcrs = Teuchos::rcp(new tpetra_multivector_type(A_bcrs->getRangeMap(),  nrhs));
+
+    // matrix vector multiplication
+    {
+      for (LO iter=0;iter<repeat;++iter) {
+        RCP<TimeMonitor> timerBlockCrsApply 
+          = rcp(new TimeMonitor(*TimeMonitor::getNewTimer("5) BlockCrs Apply")));        
+        A_bcrs->apply(*X, *B_bcrs);
+      }
+    }
+      
+    // direct conversion: block crs -> point crs 
+    RCP<tpetra_crs_matrix_type> A_crs;
+    {
+      RCP<TimeMonitor> timerConvertBlockCrsToPointCrs 
+        = rcp(new TimeMonitor(*TimeMonitor::getNewTimer("6) Conversion from BlockCrs to PointCrs")));
+
+      // construct row mapd and column map
+      // row map can be obtained from A_bcrs->getDomainMap();
+      // we anyway need both and Tpetra does not provide an interface to contruct  
+      // crs matrix without column map
+      decltype(mesh_gids) crs_gids("crs_gids", mesh_gids.extent(0)*blocksize);
+      Kokkos::parallel_for
+        (num_owned_and_remote_elements,
+         KOKKOS_LAMBDA(const LO idx) {
+          for (LO l=0;l<blocksize;++l) 
+            crs_gids(idx*blocksize+l) = mesh_gids(idx)*blocksize+l;
+        });
+      const auto owned_crs_gids = Kokkos::subview(crs_gids, local_ordinal_range_type(0, num_owned_elements*blocksize));
+      
+      RCP<const map_type> row_crs_map = rcp(new map_type(TpetraComputeGlobalNumElements, owned_crs_gids, 0, comm));
```

Comment:
BlockMultiVector should have methods for creating these two (point) Maps.  In fact, it can create the Maps for you.

Reply:
What is the right way to use multivectors with a block crs ? Intuitively, I also thought about block multi vector. Is the interface general and performant enough to use the object with other solver packages (muelu, belos and ifpack2) ?